src/terminal/adapter/PageManager.cpp

```diff
@@ -0,0 +1,254 @@
+        for (auto i = 0; i < pageSize.height; i++)
+        {
+            visibleBuffer.GetMutableRowByOffset(visibleTop + i).CopyFrom(newBuffer.GetRowByOffset(i));
+        }
+        _visiblePageNumber = newPageNumber;
+        redrawRequired = true;
+    }
+
+    // If the active page was previously visible, and is now still visible,
+    // there is no need to update any buffer properties, because we'll have
+    // been using the main buffer in both cases.
+    const auto isVisible = newPageNumber == _visiblePageNumber;
+    if (!wasVisible || !isVisible)
+    {
+        // Otherwise we need to copy the properties from the old buffer to the
+        // new, so we retain the current attributes and cursor position. This
+        // is only needed if they are actually different.
+        auto& oldBuffer = wasVisible ? visibleBuffer : _getBuffer(_activePageNumber, pageSize);
+        auto& newBuffer = isVisible ? visibleBuffer : _getBuffer(newPageNumber, pageSize);
+        if (&oldBuffer != &newBuffer)
+        {
+            // When copying the cursor position, we need to adjust the y
+            // coordinate to account for scrollback in the visible buffer.
+            const auto oldTop = wasVisible ? visibleTop : 0;
+            const auto newTop = isVisible ? visibleTop : 0;
+            auto position = oldBuffer.GetCursor().GetPosition();
+            position.y = position.y - oldTop + newTop;
+            newBuffer.SetCurrentAttributes(oldBuffer.GetCurrentAttributes());
+            newBuffer.CopyProperties(oldBuffer);
+            newBuffer.GetCursor().SetPosition(position);
+        }
+        // If we moved from the visible buffer to a background buffer we need
+        // to hide the cursor in the visible buffer. This is because the page
+        // number is like a third dimension in the cursor coordinate system.
+        // If the cursor isn't on the visible page, it's the same as if its
+        // x/y coordinates are outside the visible viewport.
+        if (wasVisible && !isVisible)
+        {
+            visibleBuffer.GetCursor().SetIsVisible(false);
+        }
+    }
+
+    _activePageNumber = newPageNumber;
+    if (redrawRequired)
+    {
+        _renderer.TriggerRedrawAll();
+    }
+}
+
+void PageManager::MoveRelative(const til::CoordType pageCount, const bool makeVisible)
+{
+    MoveTo(_activePageNumber + pageCount, makeVisible);
+}
+
+void PageManager::MakeActivePageVisible()
+{
+    if (_activePageNumber != _visiblePageNumber)
+    {
+        MoveTo(_activePageNumber, true);
+    }
+}
+
+TextBuffer& PageManager::_getBuffer(const til::CoordType pageNumber, const til::size pageSize) const
+{
+    auto& buffer = til::at(_buffers, pageNumber - 1);
+    if (buffer == nullptr)
+    {
+        // Page buffers are created on demand, and are sized to match the active
+        // page dimensions without any scrollback rows.
+        buffer = std::make_unique<TextBuffer>(pageSize, TextAttribute{}, 0, false, _renderer);

```

Comment:
(BTW It'd be great if we could move to using `shared_ptr<TextBuffer>` over time so that we can more safely share buffers with the rendering thread, or simply between components.)

Reply:
OK. I just though the unique ptr was more efficient, but I'm happy to change that.